tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+    EVP_PKEY_free(pkey);
+    return NULL;
+  }
+
+  EVP_PKEY_free(pkey);
+
+  return ctx;
+}
+
+//
+// setup_private_evp_context()
+//
+EVP_PKEY_CTX *setup_private_evp_context(const char *filepath)
+{
+  FILE *f = fopen(filepath, "r");
+
+  if (NULL == f)
+  {
+    kmyth_log(LOG_ERR, "Failed to open private key file.");
+    return NULL;
+  }
+
+  EVP_PKEY *pkey = PEM_read_PrivateKey(f, NULL, NULL, NULL);
+
+  if (NULL == pkey)
+  {
+    kmyth_log(LOG_ERR, "Failed to load private key file.");
+    if (fclose(f) != 0)
+    {
+      kmyth_log(LOG_ERR, "Failed to close private key file.");
+    }
+    return NULL;
+  }
+
+  if (fclose(f) != 0)
+  {
+    kmyth_log(LOG_ERR, "Failed to close private key file.");
+    EVP_PKEY_free(pkey);
+    return NULL;
+  }
+
+  EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pkey, NULL);
+
+  if (NULL == ctx)
+  {
+    kmyth_log(LOG_ERR, "Failed to create private key EVP context.");
+    EVP_PKEY_free(pkey);
+
+    return NULL;
+  }
+
+  EVP_PKEY_free(pkey);
+
+  return ctx;
+}
+
+//
+// generate_session_key()
+//
+int generate_session_key(unsigned char *nonce_a, size_t nonce_a_len,
+                         unsigned char *nonce_b, size_t nonce_b_len,
+                         unsigned char **key, size_t *key_len)
+{
+  // Build the combined nonce as the base for key generation.
+  size_t len = nonce_a_len + nonce_b_len;
+  unsigned char *nonces = calloc(len, sizeof(unsigned char));
+  unsigned char *index = nonces;
+
+  memcpy(index, nonce_a, nonce_a_len);
+  index += nonce_a_len;
+  memcpy(index, nonce_b, nonce_b_len);
+
+  // Setup the message digest context.
+  //const EVP_MD *type = EVP_sha3_512();
+  const EVP_MD *type = EVP_shake256();
+
+  if (NULL == type)
+  {
+    //kmyth_log(LOG_ERR, "Failed to obtain the SHA512 MD.");
+    kmyth_log(LOG_ERR, "Failed to obtain the SHAKE-256 MD.");
+    free(nonces);
+    return 1;
+  }
+  EVP_MD_CTX *ctx = EVP_MD_CTX_new();
+
+  if (NULL == ctx)
+  {
+    kmyth_log(LOG_ERR, "Failed to create the MD context.");
+    free(nonces);
+    return 1;
+  }
+
+  // Initialize the context and load in the nonces to generate the session
+  // key.
+  int result = EVP_DigestInit_ex(ctx, type, NULL);
+
+  if (0 == result)
+  {
+    kmyth_log(LOG_ERR, "Failed to initialize the MD context.");
+    EVP_MD_CTX_free(ctx);
+    free(nonces);
+    return 1;
+  }
+
+  result = EVP_DigestUpdate(ctx, nonces, len);
+  if (0 == result)
+  {
+    kmyth_log(LOG_ERR, "Failed to update the MD context.");
+    EVP_MD_CTX_free(ctx);
+    free(nonces);
+    return 1;
+  }
+
+  free(nonces);
+
+  *key = calloc(EVP_MAX_MD_SIZE, sizeof(unsigned char));
+  *key_len = EVP_MAX_MD_SIZE;
+
+  result = EVP_DigestFinal_ex(ctx, *key, (unsigned int *) key_len);
+  if (0 == result)
+  {
+    kmyth_log(LOG_ERR, "Failed to finalize the MD context.");
+    EVP_MD_CTX_free(ctx);
+    free(key);
+    *key = NULL;
+    *key_len = 0;
+    return 1;
+  }
+
+  EVP_MD_CTX_free(ctx);
+
+  //if (*key_len != 64)
+  if (*key_len != 32)
```

Comment:
As with the nonce sizes, if we used a macro to specify the generated session key length, it may improve readability and/or ease of potential future configuration.

Reply:
Agreed, I've added an `NSL_SESSION_KEY_LEN` constant to represent this going forward. We can tweak as needed.